packages/material-ui/src/ButtonBase/ButtonBase.js

```diff
@@ -99,82 +121,115 @@ class ButtonBase extends React.Component {
         ].join(' '),
       );
     }
-    prepareFocusVisible(button.ownerDocument);
+    return button;
+  }, []);
+  const { isFocusVisible, onBlurVisible } = useIsFocusVisible(getOwnerDocument);
 
-    if (this.props.action) {
-      this.props.action({
+  React.useEffect(() => {
+    if (action) {
+      action({
         focusVisible: () => {
-          this.setState({ focusVisible: true });
-          this.getButtonNode().focus();
+          setFocusVisible(true);
+          buttonRef.current.focus();
         },
       });
     }
-  }
+  }, [action, buttonRef]);
 
-  componentDidUpdate(prevProps, prevState) {
-    if (
-      this.props.focusRipple &&
-      !this.props.disableRipple &&
-      !prevState.focusVisible &&
-      this.state.focusVisible
-    ) {
-      this.ripple.pulsate();
+  React.useEffect(() => {
+    if (focusVisible && focusRipple && !disableRipple) {
+      rippleRef.current.pulsate();
     }
-  }
+  }, [disableRipple, focusRipple, focusVisible]);
```

Comment:
We likely just want to pulsate when we transition from !focusVisible => focusVisible (which matches the old behavior). However, this may actually be a bugfix where a focusVisible button gets the focusRipple enabled we likely want to start pulsating?